File: 2024/cpp/src/bin/day01.cpp

#include "lib/file_utils.hpp"
#include "lib/transform.hpp"

#include <algorithm>
#include <iostream>
#include <ranges>
#include <unordered_map>
#include <vector>

/**
 * /file
 *
 * Advent of code challenge 2024.
 * Day 1: Historian Hysteria
 *
 * See <https://adventofcode.com/2024/day/1>
 */

int main(int argc, char* argv[]) {
    std::vector<std::string> arguments(argv, argv + argc);
    if (arguments.size() != 2) {
        std::filesystem::path progname(arguments[0]);
        std::cerr << "USAGE: " << progname.filename() << " <filename.dat>" << std::endl;
        return {};
    }

    const auto lines = scp::getLines(arguments[1]);
    std::vector<int> first;
    std::vector<int> second;
    size_t count = 0;
    for (auto line : lines | std::ranges::views::filter([](auto s) { return !s.empty(); })) {
        auto numbers =                                                      //
            line                                                            //
            | std::ranges::views::split(std::string(" "))                   //
            | std::ranges::views::filter([](auto s) { return !s.empty(); }) //
            | std::ranges::views::transform(scp::parseInt());

        auto numvec = std::vector(numbers.begin(), numbers.end());
        if (numvec.size() != 2) {
            std::cerr << "ERROR: invalid input at line " << count << ", read " << numvec.size()
                      << " numbers" << std::endl;
            return -1;
        }
        first.push_back(numvec[0]);
        second.push_back(numvec[1]);
        ++count;
    }

    std::ranges::sort(first);
    std::ranges::sort(second);

    int total = 0;
    for (size_t i = 0; i < first.size(); ++i) {
        total += std::abs(second[i] - first[i]);
    }
    std::cout << "part1 result: " << total << std::endl;

    std::unordered_map<int, size_t> occurences;
    std::ranges::for_each(second, [&occurences](auto val) {
        auto n = occurences[val];
        occurences[val] = n + 1;
    });
    total = 0;
    for (size_t i = 0; i < first.size(); ++i) {
        total += first[i] * occurences[first[i]];
    }
    std::cout << "part2 result: " << total << std::endl;
}

File: 2024/rust/src/bin/day01.rs
//!
//! Advent of code challenge 2024.
//! Day 1: Historian Hysteria
//!
//! See <https://adventofcode.com/2024/day/1>

use std::collections::HashMap;
use std::env;
use std::fs;

fn main() {
    let argv: Vec<String> = env::args().collect();
    if argv.len() != 2 {
        panic!("USAGE {} <filename>", argv[0]);
    }
    let contents = fs::read_to_string(&argv[1]).expect("unable to read file contents");

    let lineprocessor = |line: &str| -> (i32, i32) {
        let v = line
            .split(' ')
            .filter(|&tok| !tok.trim().is_empty())
            .map(|token| token.parse::<i32>())
            .collect::<Result<Vec<_>, _>>()
            .unwrap_or_else(|_| panic!("couldn't parse integers in {line}"));
        if v.len() != 2 {
            panic!("ERROR: invalid input at  line: {}", line);
        }
        (v[0], v[1])
    };

    let (mut first, mut second): (Vec<i32>, Vec<i32>) = contents
        .lines()
        .map(lineprocessor)
        .collect::<Vec<_>>()
        .into_iter()
        .unzip();

    first.sort();
    second.sort();

    println!(
        "part1 answer: {}",
        first
            .iter()
            .zip(second.iter())
            .map(|(lhs, rhs)| (lhs - rhs).abs())
            .sum::<i32>()
    );

    let mut occurences = HashMap::new();
    second.into_iter().for_each(|n| {
        let count = occurences.get(&n).unwrap_or(&0);
        occurences.insert(n, count + 1);
    });

    println!(
        "part2 answer: {}",
        first
            .iter()
            .map(|n| n * occurences.get(n).unwrap_or(&0))
            .sum::<i32>()
    );
}
